[MRG] Successive halving for faster parameter search (#13900)
scikit-learn/scikit-learn: doc/modules/grid_search.rst

- a cross-validation scheme; and
- a :ref:`score function <gridsearch_scoring>`.

Two generic approaches to parameter search are provided in
scikit-learn: for given values, :class:`GridSearchCV` exhaustively considers
all parameter combinations, while :class:`RandomizedSearchCV` can sample a
given number of candidates from a parameter space with a specified
distribution. Both these tools have successive halving counterparts
:class:`HalvingGridSearchCV` and :class:`HalvingRandomSearchCV`, which can be
much faster at finding a good parameter combination.

> After describing these tools we detail, :ref:`best practices

Reviewer: ```suggestion
After describing these tools we detail :ref:`best practices
```

      Random search for hyper-parameter optimization,
      The Journal of Machine Learning Research (2012)

.. _successive_halving_user_guide:

Searching optimal parameters with successive halving
====================================================

Scikit-learn also provides the :class:`GridHalvingSearchCV` and
:class:`RandomHalvingSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive
halving is an iterative selection process where all candidates are evaluated
with a small amount of resources at the first iteration. Only some of
these candidates are selected for the next iteration, which will be
allocated more resources. What defines a resource is typically the number of

> samples to train on, or the number of trees for a gradient boosting /

Reviewer: number of iterations in iterative algorithms like gradient boosting? I mean for random forests it's not iterations but for linear models and neural nets it could be.

decision forest estimator.

As illustrated in the figure below, only a small subset of candidates 'survive'
until the last iteration. These are the candidates that have consistently been
part of the best candidates across all iterations.

.. figure:: ../auto_examples/svm/images/sphx_glr_plot_successive_halving_iterations_001.png
   :target: ../auto_examples/model_selection/plot_successive_halving_iterations.html
   :align: center

The amount of resources ``r_i`` allocated for each candidate at iteration
``i`` is controlled by the parameters ``ratio`` and ``r_min`` as follows::

    r_i = ratio**i * r_min

> ``r_min`` is the amount of resources used at the first iteration and

Reviewer: Relating to the comments by @jnothman below, I think here is a good place to say how these are determined in practice and what the knobs are the user is tuning.
For example, here you could say ``r_min`` is the amount of resources used at the first iteration. When using the dataset size as the resource, the default is to set this using a heuristic using the number of classes. For example, in a 5 class classification problem using 5-fold cross-validation it doesn't make sense to train a model with less than 25 samples at a bare minimum.

Or something like that.

Author: This is detailed in the docstring. Do you want to add it again here?

``ratio`` defines the proportions of candidates that will be selected for
the next iteration::

>     n_candidates_to_keep = n_candidates_at_i // ratio

Reviewer: Ratio is usually set to 3 or 4, though the name "Successive halving" comes from using 2 in earlier works. It's usually safe to leave this at the default of 3.

So in the first iteration, we use ``r_min`` resources ``n_candidates``
times. In the second iteration, we use ``r_min * ratio`` resources
``n_candidates // ratio`` times. The third again multiplies the resources
per candidate and divides the number of candidates. This process stops when
the maximum budget per candidate is reached, or when less than ``ratio``

> candidates are left.

Reviewer: Either here, or maybe even earlier, I would talk about how the number of candidates relates to the number of iterations and give an example, maybe using the table below.
In our implementation, the number of candidates is actually inferred using r_min and the number of iterations.

So I would make the example something like
"Assume r_min is 10 and we have a dataset of 10000 samples. That allows us to run X iterations of successive halving. That means if we start with ... candidates, we will end up with n_ratio candidates exactly when we made use of the whole dataset.
So in the case of using the dataset as a budget, it makes sense to automatically determine the number of candidates. These are usually sampled randomly from a distribution of candidate parameters.
Using more candidates will result in more models being trained on the full dataset, meaning the algorithm will perform closer to standard RandomizedSearchCV, while using less candidates will result in not all of the data being used.
"

something like that?
And then go over the case of using ``n_iterations`` and ``n_estimators`` again?

Maybe it would be nice to also discuss the idea of the brackets as they are used in hyper-band. I.e. if we need a lot of resources to figure out which parameter setting is best, we need to run a lot of configurations to the end, so something like RandomizedSearchCV is better (or equivalently having a high r_min given a particular number of candidates).

If we only need very little resources to distinguish good and bad parameters, then our pretty aggressive schedule from above, that only runs ratio many configurations on the full dataset might be enough and is much faster.

      Random search for hyper-parameter optimization,
      The Journal of Machine Learning Research (2012)

.. _successive_halving_user_guide:

Searching optimal parameters with successive halving
====================================================

Scikit-learn also provides the :class:`HalvingGridSearchCV` and
:class:`HalvingRandomSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive
halving is an iterative selection process where all candidates (the
parameter combinations) are evaluated with a small amount of resources at
the first iteration. Only some of these candidates are selected for the next
iteration, which will be allocated more resources. What defines a resource
is typically the number of samples to train on, or the number of iterations

> in iterative algorithms like gradient boosting.

Reviewer 2: I think we need a full-fledged example with `HistGradientBoostingClassifier` with `warn_start=True` and `HalvingRandomSearchCV` with `resource="max_iter"` and a fixed `n_candidates=30` that would compare the results with `RandomizedSearchCV` with `n_iter=30` to see the kind of speed up we could get.

Reviewer 2: Alright. I did some tests and even without `warm_start` support it seems possible to benefit from max_iter so it would be good to have an example that compare the use of `resource='n_samples'` vs `resource='max_iter'` and discus the pros and cons of each approach (and compare to a `RandomizedSearchCV` baseline.

candidates at the last iteration. On the other hand if we start with a small
number of candidates, the last iteration might use less than 640 samples
which is a waste of resources.

In the case of :class:`HalvingGridSearchCV`, the number of candidates is set
by default such that the maximum amount of resources is used at the last
iteration.

Changing the value of ``min_resources`` will impact the number of possible
iterations, and as a result will also have an effect on the ideal number of
candidates.

Another consideration when choosing ``min_resources`` is whether or not it
is easy to discriminate between good and bad candidates with a small amount
of resources. For example, if you need a lot of samples to distinguish
between good and bad parameters, a high ``min_resources`` (possibly with the
use of ``aggressive_elimination=True``) is recommended. On the other hand if
the distinction is clear even with a small amount of samples, then a small
``min_resources`` may be preferable since it would speed up the computation.

By default, ``min_resources`` is set to a small value (see docstrings for
details) that depends on the number of folds, and the number of classes for

> classification problems. Depending on the setting, the default valueof

Reviewer 2: ```suggestion
classification problems. Depending on the setting, the default value of
```

| 48 * 2 = 96 | 4 // 2 = 2            |
+-------------+-----------------------+

Ideally, at the last iteration, ``ratio`` candidates are evaluated, and we
can pick the best one. Note that each ``resource_iter`` is a multiple of both
``ratio`` and ``min_resources``.

The amount of resource that is used at each iteration can be found using the
`cv_results_` after converting it to a dataframe:
`results.groupby('iter')['resource_iter'].unique()`

Choosing a resource to budget
-----------------------------

By default, the budget is defined in terms of number of samples. That is,
each iteration will use an increasing amount of samples to train on. You can
however manually specify a parameter to use as the budget with the
``resource`` parameter. Here is an example where the budget is defined in
terms of the number of estimators of a random forest::

    >>> from sklearn.datasets import make_classification
    >>> from sklearn.ensemble import RandomForestClassifier
    >>> from sklearn.model_selection import HalvingGridSearchCV
    >>> import pandas as pd
    >>>
    >>> param_grid = {'max_depth': [3, 5, 10],
    ...               'min_samples_split': [2, 5, 10]}

>     >>> base_estimator = RandomForestClassifier(random_state=0)

Reviewer 2: wouldn't it make sense to use `warm_start=True` to reuse the trees that were previously grown in the previous iteration since `HalvingGridSearchCV` is called with `resource='n_estimators'`?

Reviewer 2: Actually we probably need a test for this case to check that it's possible to incrementally build a random forest this way.

Reviewer 2: Ok let's differ the `warm_start=True` optimization for later then.

    >>> results.groupby('iter')['resource_iter'].unique()
    iter
    0    [20]
    1    [40]
    2    [80]
    Name: resource_iter, dtype: object

The search process will only use 80 resources at most, while our maximum
amount of available resources is ``n_samples=1000``. Note in this case that
``min_resources = r_0 = 20``. In order for the last iteration to use as many
resources as possible, you can use the ``force_exhaust_resources``
parameter.::

    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
    ...                            ratio=2, force_exhaust_resources=True,
    ...                            ).fit(X, y)
    >>> results = pd.DataFrame.from_dict(sh.cv_results_)
    >>> results.groupby('iter')['resource_iter'].unique()
    iter
    0     [250]
    1     [500]
    2    [1000]
    Name: resource_iter, dtype: object

`min_resources` was here automatically set to 250, which results in the last
iteration using all the resources. Since ``force_exhaust_resources`` chooses an
appropriate ``min_resources`` to start with, ``min_resources`` must be set
to 'auto' (default).

Aggressive elimination of candidates
------------------------------------

> Ideally, we want the last iteration to evaluate ``ratio`` candidates. We then

Reviewer 2: It's not clear why we would want to have the latest iteration to evaluate `ratio` candidates.

To me the following schedule could make perfect sense:

iteration | n_samples | n_candidate
-----------|---------------|----------------
1            | 2500          |  40         
2            | 5000          |  20             
3            | 10000        |  10               

Assuming the time complexity of each call to fit is linear in `n_samples`, each iteration of the halving search with ratio=2 would still be balanced and I would better trust the ability of the search to find the best parameters for the full training set out of 10 candidates instead of 2.

Reviewer 2: Shall we introduce an addtitional `min_n_candidates` parameter?

Author: We always end up with a single "best" candidate at the end.

The ideal case is when this single candidate is the best out of less than `ratio` candidates, but you still get one left regardless.

times. In the second iteration, we use ``min_resources * ratio`` resources
``n_candidates // ratio`` times. The third again multiplies the resources
per candidate and divides the number of candidates. This process stops when
the maximum amount of resource per candidate is reached, or when less than
``ratio`` candidates are left.

Here is an example with ``min_resources=3`` and ``ratio=2``, starting with 70
candidates:

+-------------+-----------------------+
| ``resource_iter``     | ``n_candidates_at_i`` |
+=============+=======================+
| 3 (=min_resources)  | 70 (=n_candidates)    |
+-------------+-----------------------+
| 3 * 2 = 6   | 70 // 2 = 35          |
+-------------+-----------------------+
| 6 * 2 = 12  | 35 // 2 = 17          |
+-------------+-----------------------+
| 12 * 2 = 24 | 17 // 2 = 8           |
+-------------+-----------------------+
| 24 * 2 = 48 | 8 // 2 = 4            |
+-------------+-----------------------+
| 48 * 2 = 96 | 4 // 2 = 2            |
+-------------+-----------------------+

Ideally, at the last iteration, ``ratio`` candidates are evaluated, and we
can pick the best one. Note that each ``resource_iter`` is a multiple of both

> ``ratio`` and ``min_resources``.

Author: does this help? https://github.com/scikit-learn/scikit-learn/pull/13900/files#discussion_r318068365

> ``ratio`` and ``min_resources``.

Reviewer 3: for some reason it doesn't load the comment for me, and I can't find it. Maybe mention me on that thread?

> ``ratio`` and ``min_resources``.

Author: github seems to have messed up the links https://github.com/scikit-learn/scikit-learn/pull/13900#discussion_r318068365

small with respect to the number of candidates, the last iteration may have
to evaluate more than ``ratio`` candidates::

    >>> from sklearn.datasets import make_classification
    >>> from sklearn.svm import SVC
    >>> from sklearn.model_selection import HalvingGridSearchCV
    >>> import pandas as pd
    >>>
    >>>
    >>> param_grid = {'kernel': ('linear', 'rbf'),
    ...               'C': [1, 10, 100]}
    >>> base_estimator = SVC(gamma='scale')
    >>> X, y = make_classification(n_samples=1000)
    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
    ...                          ratio=2, max_resources=40,
    ...                          aggressive_elimination=False).fit(X, y)
    >>> results = pd.DataFrame.from_dict(sh.cv_results_)
    >>> results.groupby('iter').resource_iter.unique()
    iter
    0    [20]
    1    [40]
    Name: resource_iter, dtype: object
    >>> results.groupby('iter').resource_iter.count()  # number of candidates used at each iteration
    iter
    0    6
    1    3
    Name: resource_iter, dtype: int64

Since we cannot use more than ``max_resources=40`` resources, the process has to

> stop at the second iteration which evaluates more than ``ratio=2`` candidates.

Author: marking this is resolved because I updated above the reason why we want `ratio` or less candidates at the last iteration

  details.

.. _amount_of_resource_and_number_of_candidates:

Amount of resource and number of candidates at each iteration
-------------------------------------------------------------

The amount of resources ``resource_iter`` (e.g. the number of samples)
allocated for each candidate at iteration ``i`` is controlled by the
parameters ``ratio`` and ``min_resources`` as follows::

    resource_iter = ratio**i * min_resources

``min_resources`` is the amount of resources used at the first iteration and
``ratio`` defines the proportions of candidates that will be selected for
the next iteration (``ratio`` must be greater than 1)::

    n_candidates_to_keep = n_candidates_at_i // ratio

So in the first iteration, we use ``min_resources`` resources
``n_candidates`` times. In the second iteration, we use ``min_resources *
ratio`` resources ``n_candidates // ratio`` times. The third again
multiplies the resources per candidate and divides the number of candidates.
This process stops when the maximum amount of resource per candidate is
reached, or when we have identified the best candidate. The best candidate
is identified at the iteration that is evaluating `ratio` or less candidates

> (see below).

Author: Thanks a lot for your review @adrinjalali .

I think I addressed most points. I also updated here the UG about `ratio`. I hope it's clearer now why having `ratio` or less candidates at the last iteration is what we want. LMK!

    * :ref:`sphx_glr_auto_examples_model_selection_plot_successive_halving_iterations.py`

Choosing ``min_resources`` and the number of candidates
-------------------------------------------------------

Beside ``ratio``, the two main parameters that influence the behaviour of a
successive halving search are the ``min_resources`` parameter, and the
number of candidates (or parameter combinations) that are evaluated.
``min_resources`` is the amount of resources allocated at the first
iteration for each candidate. The number of candidates is specified directly
in :class:`HalvingRandomSearchCV`, and is determined from the ``param_grid``
parameter of :class:`HalvingGridSearchCV`.

Consider a case where the resource is the number of samples, and where we
have 1000 samples. With ``min_resources=10`` and ``ratio=2`` we are able to
run 7 iterations, with the following number of samples: ``[10, 20, 40, 80,
160, 320, 640]``.

If we start with a high number of candidates, we might end up with a lot of
candidates at the last iteration. On the other hand if we start with a small
number of candidates, the last iteration might use less than 640 samples

> which is a waste of resources.

Reviewer 3: I think reading your comment and also having read the rest of the doc now I understand what you mean, but when I was reading it the first time, it confused me quite a bit.

> which is a waste of resources.

Author: OK, yeah it's always hard to come up with a doc that's self sufficient from the start, but i hear your concern. I added more details to this section, LMK what you think (not pushed yet.)

In the case of :class:`HalvingRandomSearchCV`, the number of candidates is set
by default such that the maximum amount of resources is used at the last

> iteration.

Reviewer 3: Yeah, it was kinda missing when I was reading it.

Changing the value of ``min_resources`` will impact the number of possible
iterations, and as a result will also have an effect on the ideal number of
candidates.

Another consideration when choosing ``min_resources`` is whether or not it
is easy to discriminate between good and bad candidates with a small amount
of resources. For example, if you need a lot of samples to distinguish
between good and bad parameters, a high ``min_resources`` is recommended. On
the other hand if the distinction is clear even with a small amount of
samples, then a small ``min_resources`` may be preferable since it would
speed up the computation.

Notice in the example above that the last iteration does not use the maximum
amount of resources available: 1000 samples are available, yet only 640 are
used, at most. By default, ``min_resources`` is set to a specific value such
that the last iteration uses as many samples as possible. Please see
:ref:`exhausting_the_resources` for details.

.. _amount_of_resource_and_number_of_candidates:

Amount of resource and number of candidates at each iteration
-------------------------------------------------------------

> The amount of resources ``resource_iter`` (e.g. the number of samples)

Author: So that would be 

> The amount of resources at each iteration _i.e._ ``resource_iter`` (e.g. the number of samples)  allocated for each candidate at iteration ``i`` is controlled by the
parameters ``ratio`` and ``min_resources`` as follows::

Honestly, that doesn't read well to me. But I can see how the current one can be confusing, I'll try to come up with something else

allocated for each candidate at iteration ``i`` is controlled by the
parameters ``ratio`` and ``min_resources`` as follows::

    resource_iter = ratio**i * min_resources

``min_resources`` is the amount of resources used at the first iteration and
``ratio`` defines the proportions of candidates that will be selected for

> the next iteration (``ratio`` must be greater than 1)::

Reviewer 3: the _also_ is to kinda say that `ratio` defines many independent parameters, for some reason all called the same thing.

> the next iteration (``ratio`` must be greater than 1)::

Reviewer 3: The `min_resources` seemed like it was already explained above and redundant, and somehow gave me the impression that it's related to `ratio` here, but didn't seem like it is. Or I didn't understand it right.

> the next iteration (``ratio`` must be greater than 1)::

Author: OK, I updated with "`<formula>`, **where** min_res is ... and ratio is ...."

IMO it's normal to briefly explain each term of a formula (there's only 2 terms here so it's not too verbose). Users might arrive directly on this section without having looked at the previous one.

    n_candidates_to_keep = n_candidates_at_i // ratio

So in the first iteration, we use ``min_resources`` resources
``n_candidates`` times. In the second iteration, we use ``min_resources *
ratio`` resources ``n_candidates // ratio`` times. The third again
multiplies the resources per candidate and divides the number of candidates.
This process stops when the maximum amount of resource per candidate is
reached, or when we have identified the best candidate. The best candidate

> is identified at the iteration that is evaluating `ratio` or less candidates

Reviewer 3: I know how it works, but in my mind those two are two different parameters, don't you think?

> is identified at the iteration that is evaluating `ratio` or less candidates

Author: I'm not sure what you're saying.

There's only one parameter here, and it's used to divide the number of candidates for the next iteration. We called it `ratio`, but I'm def open to other names. By definition, it's also used to multiply the number of resources: in theory that could be another value, but that's not SH anymore then.

Note that we don't use `ratio` as a stopping criteria because we *want* to, but because we **have** to:  there is no point for the SH process to continue once we have less than `ratio` candidates. What would we do at the next iteration anyway? we would only be left with 1 candidate, so evaluating this one alone is useless, right?

> is identified at the iteration that is evaluating `ratio` or less candidates

Reviewer: I think what @adrinjalali meant was that ``ratio`` has values >1, while the name suggests they are fractions, right? Maybe ``factor`` would be better? though not sure that's worth the large replace?

> is identified at the iteration that is evaluating `ratio` or less candidates

Author: `ratio` seems to bother a lot of people. Joel suggested `rate`, which I think is good. I also like `factor`. Happy to change for whatever you might prefer?

      Random search for hyper-parameter optimization,
      The Journal of Machine Learning Research (2012)

.. _successive_halving_user_guide:

Searching optimal parameters with successive halving
====================================================

Scikit-learn also provides the :class:`HalvingGridSearchCV` and
:class:`HalvingRandomSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive

> halving is an iterative selection process where all candidates (the

Reviewer 4: It reminds me of a tournament... but maybe we can't really call this `GridTournamentCV`!

Author: This is definitely a tournament, and SH is basically a genetic algorithm without the mutation part. I think we should stick with the name from the literature, but I'll add a comment about it being a tournament

      Random search for hyper-parameter optimization,
      The Journal of Machine Learning Research (2012)

.. _successive_halving_user_guide:

> Searching optimal parameters with successive halving

Reviewer 4: Should this be "Searching _for_ optimal parameters ..."

====================================================

Scikit-learn also provides the :class:`HalvingGridSearchCV` and
:class:`HalvingRandomSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive
halving (SH) is a sort of tournament between candidate parameter combinations.
SH is an iterative selection process where all candidates (the
parameter combinations) are evaluated with a small amount of resources at
the first iteration. Only some of these candidates are selected for the next
iteration, which will be allocated more resources. What defines a resource is
typically the number of samples to train on, but it can also be an arbitrary
numeric parameter such as `n_estimators` in a random forest.

As illustrated in the figure below, only a small subset of candidates 'survive'
until the last iteration. These are the candidates that have consistently been
part of the best candidates across all iterations. Each iteration is allocated

> an increasing amount of resources, here the number of samples.

Reviewer 4: ```suggestion
an increasing amount of resources per candidate, here the number of samples.
```

      Random search for hyper-parameter optimization,
      The Journal of Machine Learning Research (2012)

.. _successive_halving_user_guide:

Searching for optimal parameters with successive halving
========================================================

Scikit-learn also provides the :class:`HalvingGridSearchCV` and
:class:`HalvingRandomSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive
halving (SH) is like a tournament among candidate parameter combinations.
SH is an iterative selection process where all candidates (the
parameter combinations) are evaluated with a small amount of resources at
the first iteration. Only some of these candidates are selected for the next

> iteration, which will be allocated more resources. What defines a resource is

Reviewer: "For parameter tuning, the resource is typically"?
I don't think the current sentence reads well. Maybe "Typical examples for resources are" or something?

typically the number of samples to train on, but it can also be an arbitrary
numeric parameter such as `n_estimators` in a random forest.

As illustrated in the figure below, only a small subset of candidates
'survive' until the last iteration. These are the candidates that have

> consistently ranked among the best candidates across all iterations. Each

Reviewer: "best" is not true, I think, though I might be nitpicking here. It could have *just* made the cutoff in each round.  Maybe "was in the subset of winners" or was in the top half, though we haven't explained the halving yet here... hm

Author: I'll replace "best" by "top scoring":

> These are the candidates that have
consistently ranked among the top-scoring candidates across all iterations

LMK if that's not OK

Reviewer: sounds good.

iteration is allocated an increasing amount of resources per candidate, here
the number of samples.

.. figure:: ../auto_examples/model_selection/images/sphx_glr_plot_successive_halving_iterations_001.png
   :target: ../auto_examples/model_selection/plot_successive_halving_iterations.html
   :align: center

The ``ratio`` parameter controls the rate at which the resources grow, and
the rate at which the number of candidates decreases. In each iteration, the
number of resources per candidate is multiplied by ``ratio`` and the number
of candidates is divided by the same ratio. Along with ``resource`` and
``min_resources``, ``ratio`` is the most important parameter to control the

> search in our implementation. ``ratio`` effectively controls the number of

Reviewer: ```suggestion
search in our implementation, though a value of 3 usually works well. ``ratio`` effectively controls the number of
```
Not sure if this is true, but I don't think ratio is a parameter that's important to think about. It obviously has a big impact but if the default is ok, maybe the user doesn't need to worry much?

iterations in :class:`HalvingGridSearchCV` and the number of candidates (if
'auto') and iterations in :class:`HalvingRandomSearchCV`.

> ``aggressive_elimination=True`` can also be used if the number of available

Reviewer: Not sure if it makes sense to mention parameters without explaining them. Maybe just say "there are several more parameters that are explained in detail below"?

Author: Joel suggested having a paragraph like this one... it's hard to satisfy every reviewer when it comes to docs :p 

We do end this paragraph a few line below with

> Each parameter and their interactions are
described in more details below.

Do you think we should say that earlier?

Reviewer: don't worry about it.

    * :ref:`sphx_glr_auto_examples_model_selection_plot_successive_halving_heatmap.py`
    * :ref:`sphx_glr_auto_examples_model_selection_plot_successive_halving_iterations.py`

Choosing ``min_resources`` and the number of candidates
-------------------------------------------------------

Beside ``ratio``, the two main parameters that influence the behaviour of a
successive halving search are the ``min_resources`` parameter, and the
number of candidates (or parameter combinations) that are evaluated.
``min_resources`` is the amount of resources allocated at the first
iteration for each candidate. The number of candidates is specified directly
in :class:`HalvingRandomSearchCV`, and is determined from the ``param_grid``
parameter of :class:`HalvingGridSearchCV`.

Consider a case where the resource is the number of samples, and where we
have 1000 samples. In theory, with ``min_resources=10`` and ``ratio=2``, we
are able to run **at most** 7 iterations with the following number of
samples: ``[10, 20, 40, 80, 160, 320, 640]``.

But depending on the number of candidates, we might run less than 7
iterations: if we start with a **small** number of candidates, the last

> iteration might use less than 640 samples, which is a waste of resources. For

Reviewer: I wouldn't say it's a waste of resources. If it's iterations, it might be good to save iterations. But if we want to ensure we use the full dataset, then it might not be great.
So you could say "which means not using all the available resources (samples)"?

example if we start with 5 candidates, we only need 2 iterations: 5
candidates for the first iteration, then `5 // 2 = 2` candidates at the
second iteration, after which we know which candidate performs the best (so
we don't need a third one). We would only be using at most 20 samples which
is a waste since we have 1000 samples at our disposal.
On the other hand, if we start with a **high** number of candidates, we might

> end up with a lot of candidates at the last iteration, which is not always

Reviewer: why is it not ideal? "which means many candidates will run with the full resources, basically reducing the procedure to standard search"?

ideal.

In the case of :class:`HalvingRandomSearchCV`, the number of candidates is set

> by default such that the maximum amount of resources is used at the last

Reviewer: "such that all the available resources are used in the last iteration"? That's not really true because of rounding.
"such that the last iteration uses as much of the available resources as possible"?

iteration. For :class:`HalvingGridSearchCV`, the number of candidates is
determined by the `param_grid` parameter. Changing the value of
``min_resources`` will impact the number of possible iterations, and as a
result will also have an effect on the ideal number of candidates.

Another consideration when choosing ``min_resources`` is whether or not it
is easy to discriminate between good and bad candidates with a small amount
of resources. For example, if you need a lot of samples to distinguish
between good and bad parameters, a high ``min_resources`` is recommended. On
the other hand if the distinction is clear even with a small amount of
samples, then a small ``min_resources`` may be preferable since it would
speed up the computation.

Notice in the example above that the last iteration does not use the maximum
amount of resources available: 1000 samples are available, yet only 640 are
used, at most. By default, ``min_resources`` is set to a specific value such

> that the last iteration uses as many samples as possible. Please see

Reviewer: Maybe see if you can consolidate that with the sentence I commented on above?

70 candidates:

+-----------------------+-----------------------+
| ``resource_iter``     | ``n_candidates_at_i`` |
+=======================+=======================+
| 3 (=min_resources)    | 70 (=n_candidates)    |
+-----------------------+-----------------------+
| 3 * 2 = 6             | 70 // 2 = 35          |
+-----------------------+-----------------------+
| 6 * 2 = 12            | 35 // 2 = 17          |
+-----------------------+-----------------------+
| 12 * 2 = 24           | 17 // 2 = 8           |
+-----------------------+-----------------------+
| 24 * 2 = 48           | 8 // 2 = 4            |
+-----------------------+-----------------------+
| 48 * 2 = 96           | 4 // 2 = 2            |
+-----------------------+-----------------------+

We can note that:

- the process stops at the first iteration which evaluates `ratio=2`
  candidates: the best candidate is the best out of these 2 candidates. It
  is not necessary to run an additional iteration, since it would only
  evaluate one candidate (namely the best one, which we have already
  identified). For this reason, **in general, we want the last iteration to

>   run at most ``ratio`` candidates**.

Reviewer: I'm not sure I agree with that. Why "at most"? We get the maximum speed-up in that case but if we want to spend more resources doing it differently is fine, right?

Author: The way I understand it is that if you evaluate more than `ratio` candidates at the last iteration, then you didn't perform SH: you performed SH up to the second to last iteration, and then the last iteration was just a regular search.

I understand SH as being a tournament where each iteration should have the same rules. If you evaluate more than ratio candidates at the last iteration, that iteration has different rules because it selects the top-1 candidate instead of the top-k.

I'll remove the bold font though ^^

such that the last iteration can use as many resources as possible, within
the `max_resources` limit::

    >>> sh = HalvingGridSearchCV(base_estimator, param_grid, cv=5,
    ...                          ratio=2, min_resources='exhaust').fit(X, y)
    >>> sh.n_resources_
    [250, 500, 1000]

`min_resources` was here automatically set to 250, which results in the last
iteration using all the resources. The exact value that is used depends on
the number of candidate parameter, on `max_resources` and on `ratio`.

For :class:`HalvingRandomSearchCV`, exhausting the resources can be done in 2
ways:

- by setting `min_resources='exhaust'`, just like for
  :class:`HalvingGridSearchCV`;
- by setting `n_candidates='exhaust'`.

Both options are mutally exclusive: using `min_resources='exhaust'` requires
knowing the number of candidates, and symmetrically `n_candidates='exhaust'`
requires knowing `min_resources`.

In general, exhausting the total number of resources leads to a better final
candidate parameter, and is slightly more time-intensive.

.. _aggressive_elimination:

Aggressive elimination of candidates
------------------------------------

> Ideally, we want the last iteration to evaluate ``ratio`` candidates (see

Reviewer: To get the maximum resource reduction?

Author: I think this point isn't about maximum resource reduction but rather maximum candidate reduction as I tried to described in https://github.com/scikit-learn/scikit-learn/pull/13900/files#r483970976

Aggressive elimination is only relevant  in cases where the number of resources is actually small enough to be exhausted no matter what.

> :ref:`amount_of_resource_and_number_of_candidates`). We then just have to

Reviewer: We could always just pick the best one at any point, no matter how many points are left ;)

  18       3             1000           0.961009  {'criterion': 'entropy', 'max_depth': None, 'max_features': 9, 'min_samples_split': 10}
  19       3             1000           0.955989  {'criterion': 'gini', 'max_depth': None, 'max_features': 10, 'min_samples_split': 4}
====  ======  ===============  =================  =======================================================================================

Each row corresponds to a given parameter combination (a candidate) and a given
iteration. The iteration is given by the ``iter`` column. The ``resource_iter``
column tells you how many resources were used.

In the example above, the best parameter combination is ``{'criterion':
'entropy', 'max_depth': None, 'max_features': 9, 'min_samples_split': 10}``
since it has reached the last iteration (3) with the highest score:
0.96.

.. topic:: References:

    .. [1] K. Jamieson, A. Talwalkar,
       `Non-stochastic Best Arm Identification and Hyperparameter
       Optimization <http://proceedings.mlr.press/v51/jamieson16.html>`_, in
       proc. of Machine Learning Research, 2016.

>     .. [2] L. Li, K. Jamieson, G. DeSalvo, A. Rostamizadeh, .A Talwalkar,

Reviewer: ```suggestion
    .. [2] L. Li, K. Jamieson, G. DeSalvo, A. Rostamizadeh, A. Talwalkar,
```

Scikit-learn also provides the :class:`GridHalvingSearchCV` and
:class:`RandomHalvingSearchCV` estimators that can be used to
search a parameter space using successive halving [1]_ [2]_. Successive
halving is an iterative selection process where all candidates are evaluated
with a small amount of resources at the first iteration. Only a subset of
these candidates are selected for the next iteration, which will be
allocated more resources. What defines a resource is typically the number of
samples to train on, or the number of trees for a gradient boosting /
decision forest estimator.

As illustrated in the figure below, only a small subset of candidates 'survive'
until the last iteration. These are the candidates that have consistently been
part of the best candidates across all iterations.

.. figure:: ../auto_examples/svm/images/sphx_glr_plot_successive_halving_iterations_001.png
   :target: ../auto_examples/model_selection/plot_successive_halving_iterations.html
   :align: center

The amount of resources ``r_i`` allocated for each candidate at iteration
``i`` is controlled by the parameters ``ratio`` and ``r_min`` as follows::

>     r_i = ratio**i * r_min

Reviewer 4: I think it would be helpful to expand and explain that:
"So in the first iteration, we use `r_min` resources `n_candidates` times. In the second iteration, we use `r_min * ratio` resources `n_candidates / ratio` times. The third again multiplies the resources per candidate and divides the number of candidates. This process stops when the maximum resources[/budget] per candidate is reached."

Author: Thanks, I used your comment and also added a table to examplify

                           n_jobs=None, oob_score=False, random_state=0, verbose=0,
                           warm_start=False)

Note that it is not possible to budget on a parameter that is part of the
parameter space.

Exhausting the budget
---------------------

As mentioned above, the first iteration uses ``r_min`` resources. If you have
a big budget, this may be a waste of resource::

    >>> from sklearn.datasets import make_classification
    >>> from sklearn.svm import SVC
    >>> from sklearn.model_selection import GridSuccessiveHalving
    >>> import pandas as pd
    >>> parameters = {'kernel': ('linear', 'rbf'),
    ...               'C': [1, 10, 100]}
    >>> base_estimator = SVC(gamma='scale')
    >>> X, y = make_classification(n_samples=1000)
    >>> sh = GridSuccessiveHalving(base_estimator, parameters, cv=5,
    ...                            ratio=2).fit(X, y)
    >>> results = pd.DataFrame.from_dict(sh.cv_results_)
    >>> results.groupby('iter').r_i.unique()
    iter
    0    [20]
    1    [40]
    2    [80]
    Name: r_i, dtype: object

The search process will only use 80 resources at most, while our maximum budget

> is ``n_samples=1000``. Note in this case that ``r_min = r_0 = 20``. In order

Reviewer 5: r_min = 20...because 2 * 2 * 5, woo magic.

Author: The point is that by setting `force_exhaust_budget=True` you get a different `r_min`. I'll make this clearer.

for the last iteration to use as many resources as possible, you can use the
``force_exhaust_budget`` parameter::

    >>> sh = GridSuccessiveHalving(base_estimator, parameters, cv=5,
    ...                            ratio=2, force_exhaust_budget=True,
    ...                            ).fit(X, y)
    >>> results = pd.DataFrame.from_dict(sh.cv_results_)
    >>> results.groupby('iter').r_i.unique()
    iter

>     0     [250]

Reviewer 5: Why is this a list? Are there iterations with more than one `r_i`?

Author: That's a pandas thing.

>Are there iterations with more than one r_i

No, each iteration has exactly one r_i.

    1     [500]
    2    [1000]
    Name: r_i, dtype: object

Since ``force_exhaust_budget`` chooses an appropriate ``r_min`` to start
with, ``r_min`` must be set to 'auto'.

Aggressive elimination of candidates
------------------------------------

> Ideally, we want the last iteration to evaluate ``ratio`` candidates. We then

Reviewer 5: If `ratio==3`, this makes it seem like there are only 3 candidates.

Author: Sorry I don't understand?

with a small amount of resources at the first iteration. Only a subset of
these candidates are selected for the next iteration, which will be
allocated more resources. What defines a resource is typically the number of
samples to train on, or the number of trees for a gradient boosting /
decision forest estimator.

As illustrated in the figure below, only a small subset of candidates 'survive'
until the last iteration. These are the candidates that have consistently been
part of the best candidates across all iterations.

.. figure:: ../auto_examples/svm/images/sphx_glr_plot_successive_halving_iterations_001.png
   :target: ../auto_examples/model_selection/plot_successive_halving_iterations.html
   :align: center

The amount of resources ``r_i`` allocated for each candidate at iteration
``i`` is controlled by the parameters ``ratio`` and ``r_min`` as follows::

    r_i = ratio**i * r_min

``r_min`` is the amount of resources used at the first iteration and
``ratio`` defines the proportions of candidates that will be selected for
the next iteration::

    n_candidates_to_keep = n_candidates_at_i // ratio

Note that each ``r_i`` is a multiple of both ``ratio`` and ``r_min``.

> 

Author: It says just above:

>The amount of resources ``r_i`` allocated for each candidate at iteration
``i`` is controlled by the parameters ``ratio`` and ``r_min`` as follows...

I'm open to suggestions to make this clearer.

----

>What parameter should the user usually specify, if any?

That's precisely what I'm trying to explain in the user guide.

> 

Reviewer 4: Are the default values for `r_min` good choices?

> 

Author: Not always. I don't think we can have a good general default. I added a subsection in the doc to explain the effects of r_min.

> Choosing the budget

Author: what would you propose?

> Choosing the budget

Reviewer 4: "Choosing a resource to budget"?????? "Alternatives to budgeting by samples"?

:func:`sklearn.metrics.r2_score` for regression.  For some applications,
other scoring functions are better suited (for example in unbalanced
classification, the accuracy score is often uninformative). An alternative
scoring function can be specified via the ``scoring`` parameter of most
parameter search tools. See :ref:`scoring_parameter` for more details.

.. _multimetric_grid_search:

Specifying multiple metrics for evaluation
------------------------------------------

> :class:`GridSearchCV` and :class:`RandomizedSearchCV` allow specifying

Reviewer: And the halving ones don't? Maybe make that explicit then?

Parallelism
-----------

The parameter search tools evaluate each parameter setting independently.

> Computations can be run in parallel if your OS supports it, by using the

Reviewer: Is there an OS that doesn't support it? (I know that phrase was there before).

> keyword ``n_jobs=-1``. See function signature for more details.

Reviewer: link to the glossary?

Reviewer: FYI I think this user guide is really really good. This is such an annoyingly difficult thing to describe and I think you're doing a great job! I'm mostly nitpicking.

Author: Thanks! :)
